File: generator/src/java/darkness/generator/api/BulbGroup.java
package darkness.generator.api;

import java.awt.Color;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;
import java.util.function.Consumer;
import java.lang.IllegalArgumentException;

/**
 * A group of one or more bulbs, that can mostly be treated as one "big" bulb through the {@link BulbSet} interface.
 */
public class BulbGroup implements BulbSet, Iterable<BulbRGB> {
	private final List<BulbRGB> bulbs;
	public final int numBulbs;

	/**
	 * Create a bulb group that consists of the given bulbs.
	 */
	public BulbGroup(BulbRGB... bulbs) {
		if (bulbs == null || bulbs.length == 0) {
			throw new IllegalArgumentException("bulbs must be non-null and contain elements");
		}
		for (int i = 0; i < bulbs.length; i++) {
			if (bulbs[i] == null) {
				throw new IllegalArgumentException("The bulb at index " + i + " is null");
			}
		}
		this.bulbs = Arrays.asList(bulbs);
		this.numBulbs = bulbs.length;
	}

	/**
	 * Get a single bulb
	 * @param idx Bulb index
	 * @return
	 */
	public BulbRGB getBulb(int idx) {
		return bulbs.get(idx);
	}

	/**
	 * Set all bulbs in the group to the given RGB color.
	 */
	@Override
	public void set(int red, int green, int blue) {
		for (BulbRGB bulb : bulbs) {
			bulb.set(red, green, blue);
		}
	}

	/**
	 * Set all bulbs in the group to the given RGB color.
	 */
	@Override
	public void set(Color color) {
		for (BulbRGB bulb : bulbs) {
			bulb.set(color);
		}
	}

	/**
	 * Set all bulbs in the group to the given HSB color.
	 */
	@Override
	public void setHSB(float hue, float saturation, float brightness) {
		for (BulbRGB bulb : bulbs) {
			bulb.setHSB(hue, saturation, brightness);
		}
	}

	/**
	 * @return The red component of the first bulb in the group.
	 */
	@Override
	public int getRed() {
		return bulbs.get(0).getRed();
	}

	/**
	 * @return The green component of the first bulb in the group.
	 */
	@Override
	public int getGreen() {
		return bulbs.get(0).getGreen();
	}

	/**
	 * @return The blue component of the first bulb in the group.
	 */
	@Override
	public int getBlue() {
		return bulbs.get(0).getBlue();
	}

	/**
	 * @return A string describing the channels and current values of the bulbs in this group.
	 */
	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder("BulbSet{");
		for (BulbRGB bulb : bulbs) {
			sb.append(bulb);
			sb.append(',');
		}
		sb.setCharAt(sb.length() - 1, '}'); // Overwrite trailing comma
		return sb.toString();
	}

	/**
	 * @return The RGB color of the first bulb in the group.
	 */
	@Override
	public Color getColor() {
		return bulbs.get(0).getColor();
	}

	@Override
	public Iterator<BulbRGB> iterator() {
		return bulbs.iterator();
	}

	@Override
	public void forEach(Consumer<? super BulbRGB> action) {
		bulbs.forEach(action);
	}

	@Override
	public Spliterator<BulbRGB> spliterator() {
		return bulbs.spliterator();
	}
}

File: generator/src/kotlin/darkness/generator/api/BulbGroup.kt
package darkness.generator.api

import java.awt.Color
import java.util.*
import java.util.function.Consumer
import java.lang.IllegalArgumentException

/**
 * A group of one or more bulbs, that can mostly be treated as one "big" bulb through the [BulbSet] interface.
 *
 * Constructor creates a bulb group that consists of the given bulbs.
 */
class BulbGroup(val allBulbs: List<BulbRGB>) : BulbSet, Iterable<BulbRGB> {
    val numBulbs: Int

    constructor(vararg bulbs: BulbRGB) : this(bulbs.toList())

    /** The red component of the first bulb in the group. */
    override val red: Int
        get() = allBulbs[0].red

    /** The green component of the first bulb in the group. */
    override val green: Int
        get() = allBulbs[0].green

    /** The blue component of the first bulb in the group. */
    override val blue: Int
        get() = allBulbs[0].blue

    /** The RGB color of the first bulb in the group. */
    override val color: Color
        get() = allBulbs[0].color

    override val position: FloatArray
        get() {
            val averagePosition = floatArrayOf(0.0f, 0.0f, 0.0f)
            var count = 0
            for (bulb in allBulbs) {
                val pos = bulb.position
                averagePosition[0] += pos[0]
                averagePosition[1] += pos[1]
                averagePosition[2] += pos[2]
                count++
            }

            averagePosition[0] /= count.toFloat()
            averagePosition[1] /= count.toFloat()
            averagePosition[2] /= count.toFloat()
            return averagePosition

        }

    init {
        if (allBulbs.isEmpty()) {
            throw IllegalArgumentException("bulbs must be non-null and contain elements")
        }
        this.numBulbs = allBulbs.size
    }

    /**
     * Get a single bulb
     * @param idx Bulb index
     * @return
     */
    fun getBulb(idx: Int): BulbRGB {
        return allBulbs[idx]
    }

    /**
     * Set all bulbs in the group to the given RGB color.
     */
    override fun set(red: Int, green: Int, blue: Int, setter: ScriptBase) {
        for (bulb in allBulbs) {
            bulb[red, green, blue] = setter
        }
    }

    /**
     * Set all bulbs in the group to the given RGB color.
     */
    override fun set(color: Color, setter: ScriptBase) {
        for (bulb in allBulbs) {
            bulb[color] = setter
        }
    }

    /**
     * Set all bulbs in the group to the given HSB color.
     */
    override fun setHSB(hue: Float, saturation: Float, brightness: Float, setter: ScriptBase) {
        for (bulb in allBulbs) {
            bulb.setHSB(hue, saturation, brightness, setter)
        }
    }

    override fun relinquish(setter: ScriptBase) {
        for (bulb in allBulbs) {
            bulb.relinquish(setter)
        }
    }

    /**
     * @return A string describing the channels and current values of the bulbs in this group.
     */
    override fun toString(): String {
        val sb = StringBuilder("BulbSet{")
        for (bulb in allBulbs) {
            sb.append(bulb)
            sb.append(',')
        }
        sb.setCharAt(sb.length - 1, '}') // Overwrite trailing comma
        return sb.toString()
    }

    override fun iterator(): Iterator<BulbRGB> {
        return allBulbs.iterator()
    }

    override fun forEach(action: Consumer<in BulbRGB>) {
        allBulbs.forEach(action)
    }

    override fun spliterator(): Spliterator<BulbRGB> {
        return allBulbs.spliterator()
    }
}
